File: src/main/java/seremis/geninfusion/gui/util/GuiRectangle.java
package seremis.geninfusion.gui.util;

import seremis.geninfusion.gui.GIGui;

import java.util.Arrays;

public class GuiRectangle {

    public int x;
    public int y;
    public int w;
    public int h;
    public int u;
    public int v;

    public GuiRectangle(int x, int y, int w, int h, int u, int v) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
        this.u = u;
        this.v = v;
    }

    public boolean inRect(GIGui gui, int mouseX, int mouseY) {
        mouseX -= gui.getLeft();
        mouseY -= gui.getTop();

        return isInBetween(x, x + w, mouseX) && isInBetween(y, y + h, mouseY);
    }

    public boolean isInBetween(int x, int y, int i) {
        return i >= Math.min(x, y) && i <= Math.max(x, y);
    }

    public void setX(int x) {
        this.x = x;
    }

    public void setY(int y) {
        this.y = y;
    }

    public void setWidth(int w) {
        this.w = w;
    }

    public void setHeight(int h) {
        this.h = h;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public int getWidth() {
        return w;
    }

    public int getHeight() {
        return h;
    }

    public void draw(GIGui gui) {
        gui.drawTexturedModalRect(x, y, u, v, w, h);
    }

    public void drawString(GIGui gui, int mouseX, int mouseY, String str) {
        if(inRect(gui, mouseX, mouseY)) {
            gui.drawHoveringString(Arrays.asList(str.split("\n")), mouseX - gui.getLeft(), mouseY - gui.getTop());
        }
    }

    @Override
    public String toString() {
        return "GuiRectangle[x: " + x + ", y: " + y + ", width: " + w + ", height : " + h + "]";
    }
}

File: src/main/scala/seremis/geninfusion/gui/util/GuiRectangle.scala
package seremis.geninfusion.gui.util

import seremis.geninfusion.gui.GIGui

class GuiRectangle(var x: Int, var y: Int, var w: Int, var h: Int, var u: Int, var v: Int) {

    def inRect(gui: GIGui, mouseX: Int, mouseY: Int): Boolean = isInBetween(x, x + w, mouseX - gui.getLeft) && isInBetween(y, y + h, mouseY - gui.getTop)

    def isInBetween(x: Int, y: Int, i: Int): Boolean = i >= Math.min(x, y) && i <= Math.max(x, y)


    def setX(x: Int) {
        this.x = x
    }

    def setY(y: Int) {
        this.y = y
    }

    def setWidth(w: Int) {
        this.w = w
    }

    def setHeight(h: Int) {
        this.h = h
    }

    def getX: Int = x

    def getY: Int = y

    def getWidth: Int = w

    def getHeight: Int = h

    def draw(gui: GIGui) {
        gui.drawTexturedModalRect(x, y, u, v, w, h)
    }

    def drawString(gui: GIGui, mouseX: Int, mouseY: Int, str: String) {
        if (inRect(gui, mouseX, mouseY)) {
            gui.drawHoveringString(List(str.split("\n"):_*), mouseX - gui.getLeft, mouseY - gui.getTop)
        }
    }

    override def toString: String = "GuiRectangle[x: " + x + ", y: " + y + ", width: " + w + ", height : " + h + "]"
}